From d352609862b1925a5cb34580738c7702fdfca3d4 Mon Sep 17 00:00:00 2001
From: beviu <contact@beviu.com>
Date: Fri, 12 Dec 2025 15:30:21 +0100
Subject: [PATCH 15/17] Switch processor when EXTMEM_SWITCH_PROCESSOR is set to
 1

Switch processor before the first user fault system call when
EXTMEM_SWITCH_PROCESSOR is set to 1. This lets lets us boot a Gem5
system with KVM and switch to the emulation at the last moment to save
time.

The Gem5 configuration needs to be modifying accordingly to handle the
hypercall #8 by switching processor for this to work.
---
 src/Makefile |  2 +-
 src/core.c   | 12 ++++++++++++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/src/Makefile b/src/Makefile
index 724e656..7ff743b 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -3,7 +3,7 @@ CFLAGS = -g -Wall -O3 -fPIC -fno-omit-frame-pointer
 #CFLAGS = -g3 -Wall -O0 -fPIC
 LDFLAGS = -shared -fno-omit-frame-pointer
 INCLUDES = -I../linux/usr/include/
-LIBS = -lm -lpthread -luring -ldl -lsyscall_intercept
+LIBS = -lm -lpthread -luring -ldl -lsyscall_intercept -lm5
 
 default: libextmem-default.so
 
diff --git a/src/core.c b/src/core.c
index 369ef86..a3516f4 100644
--- a/src/core.c
+++ b/src/core.c
@@ -33,6 +33,8 @@
 #include "storagemanager.h"
 #endif
 
+#include <gem5/m5ops.h>
+#include <m5_mmap.h>
 
 pthread_t fault_thread;
 
@@ -311,6 +313,16 @@ void extmem_init()
   uswap_init_iouring();
 #endif
 
+  const char *switch_processor = getenv("EXTMEM_SWITCH_PROCESSOR");
+  if (switch_processor && strcmp(switch_processor, "1") == 0) {
+    LOG("extmem_init: switching processor\n");
+    map_m5_mem();
+    m5_hypercall_addr(8);
+
+    /* HACK: Hypercalls seems to be handled asynchronously? Without this,
+             somehow we manage to continue executing in KVM mode. */
+    usleep(1000000);
+  }
   
   #ifdef USWAP_SIGNAL  // for self-paging mode
   LOG("extmem_init: registering SIGBUS handler\n");
-- 
2.52.0

